### PR TITLE
Add code in Ruby GC to give contextual info on parent object for rare T_NONE errors

### PR DESCRIPTION
**Pull request by aadi-stripe**
Adds logging about the "parent object" surrounding a `try to mark T_NONE` error in the Ruby GC.

### Motivation

At Stripe, we are occasionally seeing `try to mark T_NONE` errors in GC which are leading to crashes -- this is a pathological, hard-to-debug case which is probably triggered by a bad interaction with some gem extension, however, we are not sure. Adding more contextual information about the parent object currently being marked will likely help us debug these issues better.

### Test plan

* Tested by calling the new code paths in some test scripts (outside of `rb_bug`, ie just printing to stdout).
* Prints information like:
```
P:T_HASH[ k:T_OBJECT(BClass),v:T_OBJECT(AClass), ]
P:T_HASH[ k:T_OBJECT(BClass),v:T_OBJECT(AClass), ]
```
when the parent object is a hash, or
```
P:T_CLASS(AClass)
```
when the parent object is a class.

* CI builds at Stripe.

See included automated tests.

### CODE REVIEW

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_PENUIA0F4GhfKY
→ https://go/builds/bui_PENUtCGqfBr56p
→ https://go/builds/bui_PENU05bcedFwbn

File: third_party/ruby/t-none-context.patch [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
++    char *buff = debug_args->buff;
++    int buff_size = debug_args->buff_size;
++    int *pos = debug_args->pos;
++
++    VALUE k_obj = (VALUE)key;
++    VALUE v_obj = (VALUE)value;
++
++    // Skip the entry that has T_NONEs, it probably caused the error condition.
++    // We care about surrounding context.
++    if (UNLIKELY(RB_TYPE_P(k_obj, T_NONE) || RB_TYPE_P(k_obj, T_NONE))) {
```
Comment (neilparikh) [inline third_party/ruby/t-none-context.patch:40]: Why `RB_TYPE_P(k_obj, T_NONE) || RB_TYPE_P(k_obj, T_NONE)` instead of `RB_TYPE_P(k_obj, T_NONE)`?

Comment (aadi-stripe) [reply third_party/ruby/t-none-context.patch:40]: Ah whoops, I meant v_obj, nice catch

Comment (aadi-stripe) [reply third_party/ruby/t-none-context.patch]: Fixed

File: third_party/ruby/t-none-context.patch [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
++    }
++
++    if (is_markable_object(objspace, v_obj)) {
++        if (BUILTIN_TYPE(v_obj) == T_OBJECT && !internal_object_p(v_obj) && RBASIC(v_obj)->klass && RTEST(RBASIC(v_obj)->klass)) {
++            // If the value is a "real" object, add information about its class
++            VALUE class_path = rb_class_path_cached(RBASIC(v_obj)->klass);
++            if (!NIL_P(class_path)) {
++                APPENDF((BUFF_ARGS, "v:T_OBJECT(%s),", RSTRING_PTR(class_path)));
++            } else {
++                APPENDF((BUFF_ARGS, "k:T_OBJECT,"));
```
Comment (neilparikh) [inline third_party/ruby/t-none-context.patch:65]:
```suggestion
+                APPENDF((BUFF_ARGS, "v:T_OBJECT,"));
```

Comment (aadi-stripe) [reply third_party/ruby/t-none-context.patch]: Fixed

**Review by jez — APPROVED** [review]

**Pull request merged**